### PR TITLE
Support dynamic imports in the node bundle

### PR DESCRIPTION
**Pull request by Tyriar**
Part of #11996

---

⚠️ This will need some testing by people more familiar with the extension to make sure web and the dynamic imported bits are working correctly.

Summary:

- `await import(...)` syntax now works in the node bundle and are put in the same bundle for web
- New dynamically imported modules:
  - portfinder
  - tcp-port-used
  - svg-to-pdfkit
  - @vscode/extension-telemetry
- extension.node.js' webpack config now correctly uses tsconfg.extension.node.js
- webpack.extension.web.config.js fallbacks didn't seem to be used so were removed
- Simplified `stripVTControlCharacters` usage in gulpfile
- New npm scripts: `webpack-analyze:node`, `webpack-analyze:web` - plus this now works for the node bundle 🎉 
- Centralized importing of reflect-metadata into `platform/ioc/reflectMetadata.ts`
- Tightened up some eslint rules
- Replaced `import x = require(...)` with `import x from ...`
- `LinkedMap` was removed as it was unused and was causing invalid JS to be output 🤷 

---

Unfortunately I can't get the stats for the previous state, but the size of the minified/production bundle before after building `main` just now was 3.72 MB, so it went down 0.85 MB

New state (stat):

All (7.1 MB)
extension.node.js (5.9 MB)
vendors-node_modules_vscode_extension-telemetry_lib_telemetryReporter_node_min_js.node.js (329.99 KB)
vendors-node_modules_microsoft_1ds-core-js_dist-esm_src_Utils_js-node_modules_microsoft_appli-f0ebb4.node.js (228.18 KB)
vendors-node_modules_portfinder_lib_portfinder_js.node.js (214.31 KB)
vendors-node_modules_microsoft_1ds-post-js_dist-esm_src_Index_js.node.js (141.91 KB)
vendors-node_modules_svg-to-pdfkit_source_js.node.js (121.94 KB)
vendors-node_modules_microsoft_1ds-core-js_dist-esm_src_Index_js.node.js (113.99 KB)
vendors-node_modules_tcp-port-used_index_js.node.js (74.02 KB)
node_modules_vscode_extension-telemetry_lib_sync_recursive_lib_Connection_-node_modules_vscod-7cf308.node.js (320 B)

image

New state (parsed):

All (3.38 MB)
extension.node.js (2.87 MB)
vendors-node_modules_vscode_extension-telemetry_lib_telemetryReporter_node_min_js.node.js (319.98 KB)
vendors-node_modules_svg-to-pdfkit_source_js.node.js (59.59 KB)
vendors-node_modules_microsoft_1ds-core-js_dist-esm_src_Utils_js-node_modules_microsoft_appli-f0ebb4.node.js (39.78 KB)
vendors-node_modules_portfinder_lib_portfinder_js.node.js (34.06 KB)
vendors-node_modules_microsoft_1ds-core-js_dist-esm_src_Index_js.node.js (25.01 KB)
vendors-node_modules_tcp-port-used_index_js.node.js (23.76 KB)
vendors-node_modules_microsoft_1ds-post-js_dist-esm_src_Index_js.node.js (22.92 KB)
node_modules_vscode_extension-telemetry_lib_sync_recursive_lib_Connection_-node_modules_vscod-7cf308.node.js (689 B)

image

### CODE REVIEW

File: src/extension.node.ts [hunk 1]
```diff
@@ -3,13 +3,8 @@
...
 
-/* eslint-disable @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports */
-
-// This line should always be right on top.
-/* eslint-disable @typescript-eslint/no-explicit-any */
-if ((Reflect as any).metadata === undefined) {
-    require('reflect-metadata');
-}
+// reflect-metadata is needed by inversify, this must come before any inversify references
+import 'reflect-metadata';
```
Comment (DonJayamanne) [inline src/extension.node.ts:7]: We cannot do this, its possible that another extension also uses `reflect-metadata` as a result of importing this again the module will blow away all of the previous data (causing other extension that use `reflect-metadta` to fail.

Hence we need to check if the global variable already exists.

Comment (DonJayamanne) [reply src/extension.node.ts:7]: I.e. it must be conditionally imported

File: src/extension.web.ts [hunk 2]
```diff
@@ -3,19 +3,14 @@
...
 
-/* eslint-disable @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports */
-
-// This line should always be right on top.
-/* eslint-disable @typescript-eslint/no-explicit-any */
-if ((Reflect as any).metadata === undefined) {
-    require('reflect-metadata');
-}
+// reflect-metadata is needed by inversify, this must come before any inversify references
+import 'reflect-metadata';
```
Comment (DonJayamanne) [inline src/extension.web.ts:7]: Same coment as earlier

Comment (DonJayamanne) [reply src/extension.web.ts:7]: In hindsight I think we should have added comments around the conditional statement

**Review by DonJayamanne — COMMENTED** [review]
Minor tweek to importing reflect-metadata
Else all looks good

File: src/extension.node.ts (hunk 1 above)
Comment (rebornix) [reply src/extension.node.ts:7]: TIL https://github.com/inversify/InversifyJS/issues/737

**Comment by Tyriar** [comment]
Curious how this error shows up for the compile task and npm webpack, but not the editor's TS integration. Maybe because of the multiple tsconfigs?

```
❯ yarn webpack
yarn run v1.22.17
warning jupyter@2022.11.100: The engine "vscode" appears to be invalid.
$ webpack
assets by status 0 bytes [cached] 1 asset

WARNING in configuration
The 'mode' option has not been set, webpack will fallback to 'production' for this value.
Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/configuration/mode/

ERROR in main
Module not found: Error: Can't resolve './src' in 'C:\Github\microsoft\vscode-jupyter'
resolve './src' in 'C:\Github\microsoft\vscode-jupyter'
  using description file: C:\Github\microsoft\vscode-jupyter\package.json (relative path: .)
    Field 'browser' doesn't contain a valid alias configuration
    using description file: C:\Github\microsoft\vscode-jupyter\package.json (relative path: ./src)
      no extension
        Field 'browser' doesn't contain a valid alias configuration
        C:\Github\microsoft\vscode-jupyter\src is not a file
      .js
        Field 'browser' doesn't contain a valid alias configuration
        C:\Github\microsoft\vscode-jupyter\src.js doesn't exist
      .json
        Field 'browser' doesn't contain a valid alias configuration
        C:\Github\microsoft\vscode-jupyter\src.json doesn't exist
      .wasm
        Field 'browser' doesn't contain a valid alias configuration
        C:\Github\microsoft\vscode-jupyter\src.wasm doesn't exist
      as directory
        existing directory C:\Github\microsoft\vscode-jupyter\src
          using description file: C:\Github\microsoft\vscode-jupyter\package.json (relative path: ./src)
            using path: C:\Github\microsoft\vscode-jupyter\src\index
              using description file: C:\Github\microsoft\vscode-jupyter\package.json (relative path: ./src/index)
                no extension
                  Field 'browser' doesn't contain a valid alias configuration
                  C:\Github\microsoft\vscode-jupyter\src\index doesn't exist
                .js
                  Field 'browser' doesn't contain a valid alias configuration
                  C:\Github\microsoft\vscode-jupyter\src\index.js doesn't exist
                .json
                  Field 'browser' doesn't contain a valid alias configuration
                  C:\Github\microsoft\vscode-jupyter\src\index.json doesn't exist
                .wasm
                  Field 'browser' doesn't contain a valid alias configuration
                  C:\Github\microsoft\vscode-jupyter\src\index.wasm doesn't exist

webpack 5.71.0 compiled with 1 error and 1 warning in 129 ms
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

vscode-jupyter   tyriar/tsconfig_cleanup__dimport  $  is  v2022.11.100 via .NET 🎯 netcoreapp2.0 via  v16.18.1 via 🐍 v3.9.13 (.venv) 
❯ yarn webpack
yarn run v1.22.17
warning jupyter@2022.11.100: The engine "vscode" appears to be invalid.
$ webpack
assets by status 0 bytes [cached] 1 asset

WARNING in configuration
The 'mode' option has not been set, webpack will fallback to 'production' for this value.
Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/configuration/mode/

ERROR in main
Module not found: Error: Can't resolve './src' in 'C:\Github\microsoft\vscode-jupyter'
resolve './src' in 'C:\Github\microsoft\vscode-jupyter'
  using description file: C:\Github\microsoft\vscode-jupyter\package.json (relative path: .)
    Field 'browser' doesn't contain a valid alias configuration
    using description file: C:\Github\microsoft\vscode-jupyter\package.json (relative path: ./src)
      no extension
        Field 'browser' doesn't contain a valid alias configuration
        C:\Github\microsoft\vscode-jupyter\src is not a file
      .js
        Field 'browser' doesn't contain a valid alias configuration
        C:\Github\microsoft\vscode-jupyter\src.js doesn't exist
      .json
        Field 'browser' doesn't contain a valid alias configuration
        C:\Github\microsoft\vscode-jupyter\src.json doesn't exist
      .wasm
        Field 'browser' doesn't contain a valid alias configuration
        C:\Github\microsoft\vscode-jupyter\src.wasm doesn't exist
      as directory
        existing directory C:\Github\microsoft\vscode-jupyter\src
          using description file: C:\Github\microsoft\vscode-jupyter\package.json (relative path: ./src)
            using path: C:\Github\microsoft\vscode-jupyter\src\index
              using description file: C:\Github\microsoft\vscode-jupyter\package.json (relative path: ./src/index)
                no extension
                  Field 'browser' doesn't contain a valid alias configuration
                  C:\Github\microsoft\vscode-jupyter\src\index doesn't exist
                .js
                  Field 'browser' doesn't contain a valid alias configuration
                  C:\Github\microsoft\vscode-jupyter\src\index.js doesn't exist
                .json
                  Field 'browser' doesn't contain a valid alias configuration
                  C:\Github\microsoft\vscode-jupyter\src\index.json doesn't exist
                .wasm
                  Field 'browser' doesn't contain a valid alias configuration
                  C:\Github\microsoft\vscode-jupyter\src\index.wasm doesn't exist

webpack 5.71.0 compiled with 1 error and 1 warning in 108 ms
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

vscode-jupyter   tyriar/tsconfig_cleanup__dimport  $  is  v2022.11.100 via .NET 🎯 netcoreapp2.0 via  v16.18.1 via 🐍 v3.9.13 (.venv) 
```

**Comment by Tyriar** [comment]
The compile differences might be vscode vs workspace `typescript` version

**Review by DonJayamanne — DISMISSED** [review]

**Comment by Tyriar** [comment]
Error in web:

```
ERROR in ./node_modules/colors/lib/system/supports-colors.js 28:9-22
Module not found: Error: Can't resolve 'os' in '/home/runner/work/vscode-jupyter/vscode-jupyter/node_modules/colors/lib/system'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "os": require.resolve("os-browserify/browser") }'
	- install 'os-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "os": false }
resolve 'os' in '/home/runner/work/vscode-jupyter/vscode-jupyter/node_modules/colors/lib/system'
  Parsed request is a module
  using description file: /home/runner/work/vscode-jupyter/vscode-jupyter/node_modules/colors/package.json (relative path: ./lib/system)
    Field 'browser' doesn't contain a valid alias configuration
    resolve as module
      /home/runner/work/vscode-jupyter/vscode-jupyter/node_modules/colors/lib/system/node_modules doesn't exist or is not a directory
      /home/runner/work/vscode-jupyter/vscode-jupyter/node_modules/colors/lib/node_modules doesn't exist or is not a directory
      /home/runner/work/vscode-jupyter/vscode-jupyter/node_modules/colors/node_modules doesn't exist or is not a directory
      /home/runner/work/vscode-jupyter/vscode-jupyter/node_modules/node_modules doesn't exist or is not a directory
      looking for modules in /home/runner/work/vscode-jupyter/vscode-jupyter/node_modules
        single file module
          using description file: /home/runner/work/vscode-jupyter/vscode-jupyter/package.json (relative path: ./node_modules/os)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              /home/runner/work/vscode-jupyter/vscode-jupyter/node_modules/os doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              /home/runner/work/vscode-jupyter/vscode-jupyter/node_modules/os.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              /home/runner/work/vscode-jupyter/vscode-jupyter/node_modules/os.js doesn't exist
        /home/runner/work/vscode-jupyter/vscode-jupyter/node_modules/os doesn't exist
      /home/runner/work/vscode-jupyter/node_modules doesn't exist or is not a directory
      /home/runner/work/node_modules doesn't exist or is not a directory
      /home/runner/node_modules doesn't exist or is not a directory
      /home/node_modules doesn't exist or is not a directory
      /node_modules doesn't exist or is not a directory
 @ ./node_modules/colors/lib/colors.js 41:23-[72](https://github.com/microsoft/vscode-jupyter/actions/runs/3528324146/jobs/5930316161#step:39:73)
 @ ./node_modules/colors/lib/index.js 1:13-32
 @ ./src/test/datascience/widgets/commUtils.ts 12:33-50
 @ ./src/test/datascience/widgets/standardWidgets.vscode.common.test.ts 36:20-42
 @ ./src/test/ sync \.common\.test$ ./datascience/widgets/standardWidgets.vscode.common.test
 @ ./src/test/web/index.ts 69:22-68
```

**Review by DonJayamanne — APPROVED** [review]

**Pull request merged**